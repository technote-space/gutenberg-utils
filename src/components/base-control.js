const { BaseControl } = wp.components;

import classnames from 'classnames';

if ( ! BaseControl.VisualLabel ) {
	BaseControl.VisualLabel = ( { className, children } ) => {
		className = classnames( 'components-base-control__label', className );
		return (
			<span className={ className }>
				{ children }
			</span>
		);
	};
}

export default BaseControl;
