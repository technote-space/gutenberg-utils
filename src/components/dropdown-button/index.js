import React from 'react';
import classnames from 'classnames';
import {Dropdown, IconButton} from '@wordpress/components';

/**
 * @param {*} icon icon
 * @param {string|*} label label
 * @param {string|*} tooltip tooltip
 * @param {string} className class name
 * @param {string} position position
 * @param {string} contentClassName content class name
 * @param {function} renderContent render content
 * @param {boolean} isActive is active?
 * @param {boolean} isDisabled is disabled?
 * @param {boolean} isDropdownDisabled is dropdown disabled?
 * @param {boolean} isHiddenIndicator is hidden indicator?
 * @returns {*} dropdown button
 * @constructor
 */
const DropdownButton = ({
  icon = 'menu',
  label,
  tooltip = label,
  className,
  position = 'top right',
  contentClassName,
  renderContent,
  isActive = false,
  isDisabled = false,
  isDropdownDisabled = false,
  isHiddenIndicator = false,
}) => {
  return <Dropdown
    className={classnames('components-dropdown-toggle', className)}
    contentClassName={classnames('components-dropdown-button__popover', contentClassName)}
    position={position}
    renderToggle={({isOpen, onToggle, onClose}) => {
      if (isOpen && isDropdownDisabled) {
        // eslint-disable-next-line no-magic-numbers
        setTimeout(onClose, 0);
      }

      return <IconButton
        className={classnames('components-dropdown-button__toggle', 'components-toolbar__control', 'utils--components-toolbar__control', {
          'is-active is-pressed': isActive,
        })}
        icon={icon}
        onClick={onToggle}
        aria-haspopup="true"
        aria-expanded={isOpen}
        aria-pressed={isActive}
        label={label}
        tooltip={tooltip}
        disabled={isDisabled}
      >
        {!isHiddenIndicator && <span className="components-dropdown-menu__indicator"/>}
      </IconButton>;
    }}
    renderContent={renderContent}
  />;
};

export default DropdownButton;
