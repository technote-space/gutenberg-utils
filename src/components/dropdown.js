/**
 * Gutenberg v6.2
 */
import Popover from './popover';

const { Component, createRef } = wp.element;

/**
 * Dropdown
 */
class Dropdown extends Component {

	constructor() {
		super( ...arguments );

		this.toggle = this.toggle.bind( this );
		this.close = this.close.bind( this );
		this.closeIfClickOutside = this.closeIfClickOutside.bind( this );

		this.containerRef = createRef();
		this.state = {
			isOpen: false,
		};
	}

	componentWillUnmount() {
		const { isOpen } = this.state;
		const { onToggle } = this.props;
		if ( isOpen && onToggle ) {
			onToggle( false );
		}
	}

	componentDidUpdate( prevProps, prevState ) {
		const { isOpen } = this.state;
		const { onToggle } = this.props;
		if ( prevState.isOpen !== isOpen && onToggle ) {
			onToggle( isOpen );
		}
	}

	toggle() {
		this.setState( state => ( {
			isOpen: ! state.isOpen,
		} ) );
	}

	/**
	 * @param {Event} event event
	 */
	closeIfClickOutside( event ) {
		if ( ! this.containerRef.current.contains( event.target ) ) {
			const target = document.querySelector( '.components-color-picker, .components-font-size-picker__dropdown-content' );
			if ( target && target.contains( event.target ) ) {
				return;
			}
			this.close();
		}
	}

	close() {
		this.setState( { isOpen: false } );
	}

	render() {
		const { isOpen } = this.state;
		const {
			renderContent,
			renderToggle,
			position = 'bottom',
			className,
			contentClassName,
			expandOnMobile,
			headerTitle,
			focusOnMount,
			popoverProps,
		} = this.props;

		const args = { isOpen, onToggle: this.toggle, onClose: this.close };

		return (
			<div className={ className } ref={ this.containerRef }>
				{ renderToggle( args ) }
				{ isOpen && (
					<Popover
						className={ contentClassName }
						position={ position }
						onClose={ this.close }
						onClickOutside={ this.closeIfClickOutside }
						expandOnMobile={ expandOnMobile }
						headerTitle={ headerTitle }
						focusOnMount={ focusOnMount }
						{ ...popoverProps }
					>
						{ renderContent( args ) }
					</Popover>
				) }
			</div>
		);
	}
}

export default Dropdown;
