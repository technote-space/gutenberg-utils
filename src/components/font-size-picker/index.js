/**
 * Gutenberg v6.1
 **/
const { map } = window.lodash;
const { __, _x, sprintf } = wp.i18n;
const { Dashicon, Button, RangeControl, NavigableMenu } = wp.components;
import Dropdown from '../dropdown';
import BaseControl from '../base-control';
import './style.scss';


/**
 * @param {*} fallbackFontSize fallback font size
 * @param {Array} fontSizes font sizes
 * @param {boolean} disableCustomFontSizes disable custom font sizes
 * @param {function} onChange on change
 * @param {*} value value
 * @param {boolean} withSlider with slider
 * @return {null|Component} font size picker
 */
function FontSizePicker({
	fallbackFontSize,
	fontSizes = [],
	disableCustomFontSizes = false,
	onChange,
	value,
	withSlider = false,
}) {
	if (disableCustomFontSizes && !fontSizes.length) {
		return null;
	}
	const onChangeValue = (event) => {
		const newValue = event.target.value;
		if (newValue === '') {
			onChange(undefined);
			return;
		}
		onChange(Number(newValue));
	};

	const currentFont = fontSizes.find((font) => font.size === value);
	const currentFontSizeName = (currentFont && currentFont.name) || (!value && _x('Normal', 'font size name')) || _x('Custom', 'font size name');

	return (
		<BaseControl>
			<BaseControl.VisualLabel>
				{__('Font Size')}
			</BaseControl.VisualLabel>
			<div className="components-font-size-picker__buttons">
				{fontSizes.length &&
				<Dropdown
					className="components-font-size-picker__dropdown"
					contentClassName="components-font-size-picker__dropdown-content"
					position="bottom"
					renderToggle={({ isOpen, onToggle }) => (
						<Button
							className="components-font-size-picker__selector"
							isLarge
							onClick={onToggle}
							aria-expanded={isOpen}
							aria-label={sprintf(
								/* translators: %s: font size name */
								__('Font size: %s'), currentFontSizeName,
							)}
						>
							{currentFontSizeName}
						</Button>
					)}
					renderContent={() => (
						<NavigableMenu>
							{map(fontSizes, ({ name, size, slug }) => {
								const isSelected = (value === size || (!value && slug === 'normal'));

								return (
									<Button
										key={slug}
										onClick={() => onChange(slug === 'normal' ? undefined : size)}
										className={`is-font-${slug}`}
										role="menuitemradio"
										aria-checked={isSelected}
									>
										{isSelected && <Dashicon icon="saved" />}
										<span className="components-font-size-picker__dropdown-text-size" style={{ fontSize: size }}>
											{name}
										</span>
									</Button>
								);
							})}
						</NavigableMenu>
					)}
				/>
				}
				{(!withSlider && !disableCustomFontSizes) &&
				<input
					className="components-range-control__number"
					type="number"
					onChange={onChangeValue}
					aria-label={__('Custom font size')}
					value={value || ''}
				/>
				}
				<Button
					className="components-color-palette__clear"
					type="button"
					disabled={value === undefined}
					onClick={() => onChange(undefined)}
					isSmall
					isDefault
				>
					{__('Reset')}
				</Button>
			</div>
			{withSlider &&
			<RangeControl
				className="components-font-size-picker__custom-input"
				label={__('Custom Size')}
				value={value || ''}
				initialPosition={fallbackFontSize}
				onChange={onChange}
				min={12}
				max={100}
				beforeIcon="editor-textcolor"
				afterIcon="editor-textcolor"
			/>
			}
		</BaseControl>
	);
}

export default FontSizePicker;
