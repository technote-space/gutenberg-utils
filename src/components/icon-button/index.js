import { comparePackageVersion } from '../../helpers/compatibility';

/**
 * @param {object} props props
 * @return {null|Component} icon button
 */
function IconButton(props) {
	let Component;

	/* istanbul ignore next */
	if (comparePackageVersion('wp-components', '8.5.0', '>=')) {
		Component = require('./implement').default;
	} else {
		/* istanbul ignore next */
		Component = wp.components.IconButton;
	}

	return (
		<Component {...props} />
	);
}

export default IconButton;
