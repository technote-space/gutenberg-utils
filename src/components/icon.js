import classnames from 'classnames';

const {SVG, Dashicon} = wp.components;
const {isURL}         = wp.url;

const {isString} = window.lodash;
import {ICON_SIZE} from '../constant';

const isSVG      = icon => isString(icon) && (isURL(icon) || /^data:image/.test(icon));
const isDashicon = icon => isString(icon) && /^dashicons-/.test(icon);

const createSVG      = (icon, className) => <SVG
  version="1.1"
  xmlns="http://www.w3.org/2000/svg"
  x="0px"
  y="0px"
  width={`${ICON_SIZE}px`}
  height={`${ICON_SIZE}px`}
  viewBox={`0 0 ${ICON_SIZE} ${ICON_SIZE}`}
  enableBackground={`new 0 0 ${ICON_SIZE} ${ICON_SIZE}`}
  className={className}
>
  <image
    width={ICON_SIZE}
    height={ICON_SIZE}
    x="0"
    y="0"
    xlinkHref={icon}
  />
</SVG>;
const createDashicon = (icon, className) => <Dashicon
  icon={icon.replace(/^dashicons-/, '')}
  className={className}
/>;

/**
 * @param {*} icon icon
 * @param {string} className class name
 * @param {*} defaultIcon default icon
 * @returns {null|*|null|*} icon
 * @constructor
 */
const Icon = (
  {
    icon,
    className,
    defaultIcon = false,
  },
) => {
  if (isSVG(icon)) {
    return createSVG(icon, classnames('components-icon components-icon__svg', className));
  }

  if (isDashicon(icon)) {
    return createDashicon(icon, classnames('components-icon components-icon__dashicon', className));
  }

  if (!icon && defaultIcon) {
    return Icon({
      icon: defaultIcon,
      className: className,
    });
  }

  if (!icon) {
    return null;
  }

  return icon;
};

export default Icon;
