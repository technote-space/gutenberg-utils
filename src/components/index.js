import Icon from './icon';
import DropdownButton from './dropdown-button';
import ColorPalette from './color-palette';
import FontSizePicker from './font-size-picker';

export { Icon, DropdownButton, ColorPalette, FontSizePicker };
