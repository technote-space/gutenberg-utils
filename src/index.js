import * as Components from './components';
import * as Helpers from './helpers';

Helpers.applyStyles('.components-dropdown-button__popover .components-font-size-picker__controls .components-custom-select-control__menu', {
	position: 'fixed',
	minWidth: '150px',
});
Helpers.applyStyles('.components-dropdown-button__popover .components-popover__content', {
	padding: '12px',
});
Helpers.applyStyles('.components-dropdown-button__has-property-color .dashicon', {
	color: '#d94f4f',
});
Helpers.applyStyles('.components-dropdown-button__has-property-background-color .dashicon', {
	backgroundColor: '#f0b849',
});
Helpers.applyStyles('.components-dropdown-toggle>.components-toolbar__control.components-button', {
	width: 'auto',
});
Helpers.applyStyles('.components-dropdown-toggle>.components-toolbar__control.components-button.has-icon.has-text>svg', {
	margin: 0,
});

export { Components, Helpers };
