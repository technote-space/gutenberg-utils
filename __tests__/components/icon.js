/* eslint-disable no-magic-numbers */
import React from 'react';
import {render} from 'enzyme';
import toJson from 'enzyme-to-json';
import {Icon as OriginalIcon} from '@wordpress/components';
import {Icon} from '../../src/components';

describe('Icon', () => {
  it('should render Icon', () => {
    const wrapper = render(
      <div>
        <Icon
          icon={'http://example.com'}
          className={'test1'}
        />
        <Icon
          icon={'data:image/png;base64,1234567890'}
          className={'test2'}
        />
        <Icon
          icon={'dashicons-menu'}
          className={'test3'}
        />
        <Icon
          icon={undefined}
          defaultIcon={'dashicons-admin-site'}
          className={'test4'}
        />
        <Icon
          icon={undefined}
          className={'test5'}
        />
        <Icon
          icon={<OriginalIcon
            icon={'admin-post'}
          />}
          className={'test6'}
        />
      </div>,
    );

    expect(toJson(wrapper, {mode: 'deep'})).toMatchSnapshot();

    expect(wrapper.find('svg')).toHaveLength(2);
    expect(wrapper.find('span')).toHaveLength(3);
    expect(wrapper.find('.test1')).toHaveLength(1);
    expect(wrapper.find('.test2')).toHaveLength(1);
    expect(wrapper.find('.test3')).toHaveLength(1);
    expect(wrapper.find('.test4')).toHaveLength(1);
    expect(wrapper.find('.test5')).toHaveLength(0);
    expect(wrapper.find('.test6')).toHaveLength(0);
    expect(wrapper.find('.dashicons-admin-post')).toHaveLength(1);
  });
});
