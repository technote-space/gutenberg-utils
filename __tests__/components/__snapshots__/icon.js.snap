// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should render Icon 1`] = `
<div>
  <svg
    aria-hidden="true"
    class="components-icon components-icon__svg test1"
    enable-background="new 0 0 20 20"
    focusable="false"
    height="20px"
    role="img"
    version="1.1"
    viewBox="0 0 20 20"
    width="20px"
    x="0px"
    xmlns="http://www.w3.org/2000/svg"
    y="0px"
  >
    <image
      height="20"
      href="http://example.com"
      width="20"
      x="0"
      y="0"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="components-icon components-icon__svg test2"
    enable-background="new 0 0 20 20"
    focusable="false"
    height="20px"
    role="img"
    version="1.1"
    viewBox="0 0 20 20"
    width="20px"
    x="0px"
    xmlns="http://www.w3.org/2000/svg"
    y="0px"
  >
    <image
      height="20"
      href="data:image/png;base64,1234567890"
      width="20"
      x="0"
      y="0"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="dashicon dashicons-menu components-icon components-icon__dashicon test3"
    focusable="false"
    height="20"
    role="img"
    viewBox="0 0 20 20"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M17 7V5H3v2h14zm0 4V9H3v2h14zm0 4v-2H3v2h14z"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="dashicon dashicons-admin-site components-icon components-icon__dashicon test4"
    focusable="false"
    height="20"
    role="img"
    viewBox="0 0 20 20"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M9 0C4.03 0 0 4.03 0 9s4.03 9 9 9 9-4.03 9-9-4.03-9-9-9zm3.46 11.95c0 1.47-.8 3.3-4.06 4.7.3-4.17-2.52-3.69-3.2-5 .126-1.1.804-2.063 1.8-2.55-1.552-.266-3-.96-4.18-2 .05.47.28.904.64 1.21-.782-.295-1.458-.817-1.94-1.5.977-3.225 3.883-5.482 7.25-5.63-.84 1.38-1.5 4.13 0 5.57C7.23 7 6.26 5 5.41 5.79c-1.13 1.06.33 2.51 3.42 3.08 3.29.59 3.66 1.58 3.63 3.08zm1.34-4c-.32-1.11.62-2.23 1.69-3.14 1.356 1.955 1.67 4.45.84 6.68-.77-1.89-2.17-2.32-2.53-3.57v.03z"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="dashicon dashicons-admin-post"
    focusable="false"
    height="20"
    role="img"
    viewBox="0 0 20 20"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M10.44 3.02l1.82-1.82 6.36 6.35-1.83 1.82c-1.05-.68-2.48-.57-3.41.36l-.75.75c-.92.93-1.04 2.35-.35 3.41l-1.83 1.82-2.41-2.41-2.8 2.79c-.42.42-3.38 2.71-3.8 2.29s1.86-3.39 2.28-3.81l2.79-2.79L4.1 9.36l1.83-1.82c1.05.69 2.48.57 3.4-.36l.75-.75c.93-.92 1.05-2.35.36-3.41z"
    />
  </svg>
</div>
`;
